File: laborations/lab4/haskell/lab4.hs
import Control.Monad        (forM_, when)
import Control.Monad.State  (StateT, execStateT, gets, modify, liftIO)
import Control.Monad.Except (runExcept)

import System.IO            (hPutStrLn, readFile, stderr)
import System.Environment   (getArgs)
import System.Exit          (exitFailure, exitSuccess)

import Fun.Abs              (Program)
import Fun.Par              (pProgram, myLexer)
import Fun.Print            (printTree)
import Fun.ErrM             (Err(..))

import Interpreter          (Strategy(..), interpret)

-- | Entry point.

main :: IO ()
main = do
  args <- getArgs
  CmdLine strategy file <- mapM_ parseArg args `execStateT` initCmdLine
  when (null file) usage
  run strategy =<< readFile file

-- | Main pipeline.

run :: Strategy -> String -> IO ()
run strategy s = eval strategy =<< parse s

-- | Parse.

parse :: String -> IO Program
parse s = case pProgram (myLexer s) of
  Bad err -> do
    putStrLn "SYNTAX ERROR"
    putStrLn err
    exitFailure
  Ok prg -> return prg


-- | Interpret in call-by-value or call-by-name.

eval :: Strategy -> Program -> IO ()
eval strategy prg = do
  case runExcept $ interpret strategy prg of
    Left err -> do
      putStrLn "INTERPRETER ERROR"
      putStrLn err
      exitFailure
    Right i -> do
      putStrLn $ show i
      hPutStrLn stderr "OK"
      exitSuccess

-- * Command-line parsing

-- | State of the command line parser.

data CmdLine = CmdLine
  { strategy :: Strategy
  , fileName :: FilePath
  }

-- | Initial state: 'CallByValue' is default.

initCmdLine :: CmdLine
initCmdLine = CmdLine
  { strategy = CallByValue
  , fileName = ""
  }

-- | Parse a single command line argument.

parseArg :: String -> StateT CmdLine IO ()
parseArg s
  | s == "-n" = modify $ \ o -> o { strategy = CallByName }
  | s == "-v" = modify $ \ o -> o { strategy = CallByValue }
  | otherwise = do
      old <- gets fileName
      if null old then modify $ \ o -> o { fileName = s }
      else liftIO $ usage

-- | Print usage information and exit.

usage :: IO ()
usage = do
  putStrLn "Usage: lab4 [-n|-v] <SourceFile>"
  exitFailure

File: laborations/lab4/java/lab4.java
import fun.*;
import fun.Absyn.*;
import java.io.*;

public class lab4 {

    public static void usage() {
        System.err.println("Usage: lab4 [-n|-v] FILE");
        System.err.println("Interprets FILE either in call-by-name (-n) or call-by-value (-v, default).");
        System.exit(1);
    }

    public static void main(String args[]) {

        // Parse command line

        String file = null;
        Strategy strategy = Strategy.CallByValue;  // call-by-value is default strategy
        for (String arg : args) {
            switch (arg) {

                case "-n":
                    strategy = Strategy.CallByName;
                    break;

                case "-v":
                    strategy = Strategy.CallByValue;
                    break;

                default:
                    if (file == null) file = arg;
                    else usage();
            }
        }
        if (file == null) usage();

        // Run interpreter

        Yylex l = null;
        try {

            l = new Yylex(new FileReader(file));
            parser p = new parser(l);
            Program prg = p.pProgram();
            new Interpreter(strategy).interpret(prg);

        } catch (RuntimeException e) {

            System.out.println("INTERPRETER ERROR");
            System.err.println(e.toString());
            System.exit(-1);

        } catch (IOException e) {

            System.err.println(e.toString());
            System.exit(1);

        } catch (Throwable e) {

            System.out.println("SYNTAX ERROR");
            System.out.println("At line " + String.valueOf(l.line_num())
                       + ", near \"" + l.buff() + "\" :");
            System.out.println("     " + e.getMessage());
            e.printStackTrace();
            System.exit(1);

        }
    }
}
